diff -puN arch/ia64/lib/kgdb_serial.c~irqaction-use-cpumask arch/ia64/lib/kgdb_serial.c
--- 25/arch/ia64/lib/kgdb_serial.c~irqaction-use-cpumask	2004-06-23 21:44:12.215208840 -0700
+++ 25-akpm/arch/ia64/lib/kgdb_serial.c	2004-06-23 21:44:12.275199720 -0700
@@ -486,7 +486,7 @@ kgdb_enable_ints_now(void)
 		irq_desc_t *desc;
 		kgdb_action.handler = gdb_interrupt;
 		kgdb_action.flags = IRQ_T(gdb_async_info);
-		kgdb_action.mask = 0;
+		kgdb_action.mask = CPU_MASK_NONE;
 		kgdb_action.name = "KGDB-stub";
 		kgdb_action.next = NULL;
 		kgdb_action.dev_id = NULL;

