

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/printk.c |    3 +++
 1 files changed, 3 insertions(+)

diff -puN kernel/printk.c~provide-console_device-comment kernel/printk.c
--- 25/kernel/printk.c~provide-console_device-comment	2004-06-27 22:24:18.939593240 -0700
+++ 25-akpm/kernel/printk.c	2004-06-27 22:24:25.478599160 -0700
@@ -683,6 +683,9 @@ void console_unblank(void)
 }
 EXPORT_SYMBOL(console_unblank);
 
+/*
+ * Return the console tty driver structure and its associated index
+ */
 struct tty_driver *console_device(int *index)
 {
 	struct console *c;
_
