

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/block/ll_rw_blk.c |    1 +
 1 files changed, 1 insertion(+)

diff -puN drivers/block/ll_rw_blk.c~unplug-can-sleep drivers/block/ll_rw_blk.c
--- 25/drivers/block/ll_rw_blk.c~unplug-can-sleep	2004-06-22 12:46:57.200472440 -0700
+++ 25-akpm/drivers/block/ll_rw_blk.c	2004-06-22 12:46:57.207471376 -0700
@@ -1205,6 +1205,7 @@ EXPORT_SYMBOL(__generic_unplug_device);
  **/
 void generic_unplug_device(request_queue_t *q)
 {
+	might_sleep();
 	spin_lock_irq(q->queue_lock);
 	__generic_unplug_device(q);
 	spin_unlock_irq(q->queue_lock);
_
