
From: "Antonino A. Daplas" <adaplas@hotpop.com>

Signed-off-by: Antonino Daplas <adaplas@pol.net>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/video/Kconfig |    7 +++++++
 1 files changed, 7 insertions(+)

diff -puN drivers/video/Kconfig~help-text-for-fb_riva_i2c drivers/video/Kconfig
--- 25/drivers/video/Kconfig~help-text-for-fb_riva_i2c	2004-06-21 00:42:35.584438856 -0700
+++ 25-akpm/drivers/video/Kconfig	2004-06-21 00:42:35.588438248 -0700
@@ -436,6 +436,13 @@ config FB_RIVA_I2C
        bool "Enable DDC Support"
        depends on FB_RIVA && I2C
        help
+	  This enables I2C support for nVidia Chipsets.  This is used
+	  only for getting EDID information from the attached display
+	  allowing for robust video mode handling and switching.
+
+	  Because fbdev-2.6 requires that drivers must be able to
+	  independently validate video mode parameters, you should say Y
+	  here.
 
 config FB_I810
 	tristate "Intel 810/815 support (EXPERIMENTAL)"
_
