---

 linux-2.6.7-xx2-xiphux/drivers/net/Kconfig |   12 ++++++++++++
 1 files changed, 12 insertions(+)

diff -puN drivers/net/Kconfig~r8169_napi-help-text drivers/net/Kconfig
--- linux-2.6.7-xx2/drivers/net/Kconfig~r8169_napi-help-text	2004-06-24 06:20:46.794559480 -0400
+++ linux-2.6.7-xx2-xiphux/drivers/net/Kconfig	2004-06-24 06:20:46.801558416 -0400
@@ -2009,7 +2009,19 @@ config R8169
 config R8169_NAPI
 	bool "Use Rx and Tx Polling (NAPI) (EXPERIMENTAL)"
 	depends on R8169 && EXPERIMENTAL
+	help
+	  NAPI is a new driver API designed to reduce CPU and interrupt load
+	  when the driver is receiving lots of packets from the card. It is
+	  still somewhat experimental and thus not yet enabled by default.
+
+	  If your estimated Rx load is 10kpps or more, or if the card will be
+	  deployed on potentially unfriendly networks (e.g. in a firewall),
+	  then say Y here.
+
+	  See <file:Documentation/networking/NAPI_HOWTO.txt> for more
+	  information.
 
+	  If in doubt, say N.
 
 config SK98LIN
 	tristate "Marvell Yukon Chipset / SysKonnect SK-98xx Support"

_
