


---

 25-akpm/fs/compat_ioctl.c |    1 +
 1 files changed, 1 insertion(+)

diff -puN fs/compat_ioctl.c~control-pktcdvd-with-an-auxiliary-character-device-fix fs/compat_ioctl.c
--- 25/fs/compat_ioctl.c~control-pktcdvd-with-an-auxiliary-character-device-fix	2004-07-27 02:08:42.642036008 -0700
+++ 25-akpm/fs/compat_ioctl.c	2004-07-27 02:08:42.648035096 -0700
@@ -114,6 +114,7 @@
 #include <linux/random.h>
 #include <linux/filter.h>
 #include <linux/msdos_fs.h>
+#include <linux/pktcdvd.h>
 
 #include <linux/hiddev.h>
 

_
