

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/ppc64/kernel/smp.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN arch/ppc64/kernel/smp.c~create-cpu_sibling_map-for-ppc64-fix arch/ppc64/kernel/smp.c
--- 25/arch/ppc64/kernel/smp.c~create-cpu_sibling_map-for-ppc64-fix	Wed Aug  4 14:52:08 2004
+++ 25-akpm/arch/ppc64/kernel/smp.c	Wed Aug  4 14:52:08 2004
@@ -61,7 +61,7 @@ cpumask_t cpu_possible_map = CPU_MASK_NO
 cpumask_t cpu_online_map = CPU_MASK_NONE;
 cpumask_t cpu_available_map = CPU_MASK_NONE;
 cpumask_t cpu_present_at_boot = CPU_MASK_NONE;
-cpumask_t cpu_sibling_map[NR_CPUS] = { [0 .. NR_CPUS-1] = CPU_MASK_NONE };
+cpumask_t cpu_sibling_map[NR_CPUS] = { [0 ... NR_CPUS-1] = CPU_MASK_NONE };
 
 EXPORT_SYMBOL(cpu_online_map);
 EXPORT_SYMBOL(cpu_possible_map);
_
