

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/video/fbmem.c |    1 +
 1 files changed, 1 insertion(+)

diff -puN drivers/video/fbmem.c~fbcon-differentiate-bits_per_pixel-from-color-depth-export drivers/video/fbmem.c
--- 25/drivers/video/fbmem.c~fbcon-differentiate-bits_per_pixel-from-color-depth-export	2004-07-31 19:02:11.591457232 -0700
+++ 25-akpm/drivers/video/fbmem.c	2004-07-31 19:02:35.826772904 -0700
@@ -444,6 +444,7 @@ int fb_get_color_depth(struct fb_info *i
 		return (var->green.length + var->red.length +
 			var->blue.length);
 }
+EXPORT_SYMBOL(fb_get_color_depth);
 
 /*
  * Drawing helpers.
_
