

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/video/console/fbcon.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN drivers/video/console/fbcon.c~fbcon-differentiate-bits_per_pixel-from-color-depth-fixup drivers/video/console/fbcon.c
--- 25/drivers/video/console/fbcon.c~fbcon-differentiate-bits_per_pixel-from-color-depth-fixup	2004-07-31 18:30:51.059341568 -0700
+++ 25-akpm/drivers/video/console/fbcon.c	2004-07-31 18:31:12.973010184 -0700
@@ -2365,7 +2365,7 @@ static int fbcon_set_palette(struct vc_d
 	 * use default palette of fbcon.
 	 */
 	} else
-		fb_copy_cmap(fb_default_cmap(1 << depth), &palette_cmap, 0);
+		fb_copy_cmap(fb_default_cmap(1 << depth), &palette_cmap);
 
 	return fb_set_cmap(&palette_cmap, info);
 }
_
