
From: Jeff Dike <jdike@addtoit.com>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/include/asm-generic/bug.h |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN include/asm-generic/bug.h~move-duplicate-bug-and-warn_on-bits-to-asm-generic-fix include/asm-generic/bug.h
--- 25/include/asm-generic/bug.h~move-duplicate-bug-and-warn_on-bits-to-asm-generic-fix	2004-08-05 20:31:46.531787264 -0700
+++ 25-akpm/include/asm-generic/bug.h	2004-08-05 20:31:46.534786808 -0700
@@ -7,7 +7,7 @@
 #ifndef HAVE_ARCH_BUG
 #define BUG() do { \
 	printk("kernel BUG at %s:%d!\n", __FILE__, __LINE__); \
-	panic("BUG!");
+	panic("BUG!"); \
 } while (0)
 #endif
 
_
