---

 25-akpm/include/linux/compiler-gcc.h |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN include/linux/compiler-gcc.h~ppc64-reloc_hide include/linux/compiler-gcc.h
--- 25/include/linux/compiler-gcc.h~ppc64-reloc_hide	Tue May 18 14:45:10 2004
+++ 25-akpm/include/linux/compiler-gcc.h	Tue May 18 14:45:10 2004
@@ -13,5 +13,5 @@
    shouldn't recognize the original var, and make assumptions about it */
 #define RELOC_HIDE(ptr, off)					\
   ({ unsigned long __ptr;					\
-    __asm__ ("" : "=g"(__ptr) : "0"(ptr));		\
+	__asm__ ("" : "=r"(__ptr) : "0"(ptr));			\
     (typeof(ptr)) (__ptr + (off)); })

_
