
From: William Lee Irwin III <wli@holomorphy.com>

sparc64 needs profile.h and to terminate comments describing %o7.

Signed-off-by: Andrew Morton <akpm@osdl.org>
Index: xx-sources/arch/sparc64/kernel/smp.c
===================================================================
--- xx-sources.orig/arch/sparc64/kernel/smp.c	2004-08-14 13:24:36.785094544 -0400
+++ xx-sources/arch/sparc64/kernel/smp.c	2004-08-14 13:28:49.224717888 -0400
@@ -19,6 +19,7 @@
 #include <linux/seq_file.h>
 #include <linux/cache.h>
 #include <linux/jiffies.h>
+#include <linux/profile.h>
 
 #include <asm/head.h>
 #include <asm/ptrace.h>
Index: xx-sources/arch/sparc64/kernel/time.c
===================================================================
--- xx-sources.orig/arch/sparc64/kernel/time.c	2004-08-14 13:27:34.252115456 -0400
+++ xx-sources/arch/sparc64/kernel/time.c	2004-08-14 13:28:49.227717432 -0400
@@ -29,6 +29,7 @@
 #include <linux/jiffies.h>
 #include <linux/cpufreq.h>
 #include <linux/percpu.h>
+#include <linux/profile.h>
 
 #include <asm/oplib.h>
 #include <asm/mostek.h>
