

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-sparc64-akpm/kernel/sched.c |    2 ++
 1 files changed, 2 insertions(+)

diff -puN kernel/sched.c~profile-tick-fix kernel/sched.c
--- 25-sparc64/kernel/sched.c~profile-tick-fix	2004-08-04 21:32:09.149862096 -0700
+++ 25-sparc64-akpm/kernel/sched.c	2004-08-04 21:32:20.932070928 -0700
@@ -43,6 +43,8 @@
 #include <linux/kthread.h>
 #include <linux/seq_file.h>
 #include <linux/times.h>
+#include <linux/profile.h>
+
 #include <asm/tlb.h>
 
 #include <asm/unistd.h>
_
