

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/sched.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN kernel/sched.c~sched-isolated-sched-domains-fix kernel/sched.c
--- 25/kernel/sched.c~sched-isolated-sched-domains-fix	2004-07-31 18:00:24.258057576 -0700
+++ 25-akpm/kernel/sched.c	2004-07-31 18:00:45.420840344 -0700
@@ -3861,7 +3861,7 @@ __init static void arch_init_sched_domai
 		if (i != first_cpu(this_sibling_map))
 			continue;
 
-		init_sched_build_groups(sched_group_cpus, this_sibling_mask,
+		init_sched_build_groups(sched_group_cpus, this_sibling_map,
 						&cpu_to_cpu_group);
 	}
 #endif
_
