

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/sched.c |    2 ++
 1 files changed, 2 insertions(+)

diff -puN kernel/sched.c~sched-domain-node-span-4-warning-fix kernel/sched.c
--- 25/kernel/sched.c~sched-domain-node-span-4-warning-fix	2004-07-27 23:36:05.409119336 -0700
+++ 25-akpm/kernel/sched.c	2004-07-27 23:36:05.415118424 -0700
@@ -3666,6 +3666,7 @@ void cpu_attach_domain(struct sched_doma
 	unlock_cpu_hotplug();
 }
 
+#ifdef CONFIG_NUMA
 #ifdef ARCH_HAS_SCHED_DOMAIN
 extern cpumask_t __init sched_domain_node_span(int node, int size);
 #else
@@ -3674,6 +3675,7 @@ static cpumask_t __init sched_domain_nod
 	return cpu_possible_map;
 }
 #endif /* ARCH_HAS_SCHED_DOMAIN */
+#endif
 
 #ifdef CONFIG_SCHED_SMT
 static DEFINE_PER_CPU(struct sched_domain, cpu_domains);
_
