

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/slab.c |    4 ++++
 1 files changed, 4 insertions(+)

diff -puN mm/slab.c~slab-latency-fix mm/slab.c
--- 25/mm/slab.c~slab-latency-fix	Tue Jul 27 14:22:08 2004
+++ 25-akpm/mm/slab.c	Tue Jul 27 14:22:08 2004
@@ -2645,6 +2645,10 @@ static void enable_cpucache (kmem_cache_
 	if (limit > 32)
 		limit = 32;
 #endif
+#ifdef CONFIG_PREEMPT
+	if (limit > 16)
+		limit = 16;
+#endif
 	err = do_tune_cpucache(cachep, limit, (limit+1)/2, shared);
 	if (err)
 		printk(KERN_ERR "enable_cpucache failed for %s, error %d.\n",
_
