

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/char/Kconfig |    1 +
 1 files changed, 1 insertion(+)

diff -puN drivers/char/Kconfig~snsc-build-fix drivers/char/Kconfig
--- 25/drivers/char/Kconfig~snsc-build-fix	2004-07-31 17:53:52.818565424 -0700
+++ 25-akpm/drivers/char/Kconfig	2004-07-31 17:54:39.658444680 -0700
@@ -426,6 +426,7 @@ config A2232
 
 config SGI_SNSC
 	bool "SGI Altix system controller communication support"
+	depends on CONFIG_IA64_SGI_SN2
 	help
 	  If you have an SGI Altix and you want to enable system
 	  controller communication from user space (you want this!),
_
