

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/thrash.c |    2 --
 1 files changed, 2 deletions(-)

diff -puN mm/thrash.c~token-based-thrashing-control-remove-debug mm/thrash.c
--- 25/mm/thrash.c~token-based-thrashing-control-remove-debug	2004-08-01 22:53:46.094375232 -0700
+++ 25-akpm/mm/thrash.c	2004-08-01 22:53:46.097374776 -0700
@@ -80,8 +80,6 @@ void grab_swap_token(void)
 			mm->swap_token_time = eligible;
 			swap_token_timeout = jiffies + SWAP_TOKEN_TIMEOUT;
 			swap_token_mm = current->mm;
-			printk("Took swap token, pid %d (%s)\n",
-				 current->pid, current->comm);
 		}
 		spin_unlock(&swap_token_lock);
 	}
_
