
From: <blaisorblade_spam@yahoo.it>

Signed-off-by: Paolo 'Blaisorblade' Giarrusso <blaisorblade_spam@yahoo.it>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/um/main.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN arch/um/main.c~uml-little-kmalloc arch/um/main.c
--- 25/arch/um/main.c~uml-little-kmalloc	2004-07-06 01:20:34.352821616 -0700
+++ 25-akpm/arch/um/main.c	2004-07-06 01:20:34.355821160 -0700
@@ -167,7 +167,7 @@ void *__wrap_malloc(int size)
 
 	if(!CAN_KMALLOC())
 		return(__real_malloc(size));
-	else if(size <= 128 * 1024) /* kmalloc is good for only 128K */
+	else if(size <= PAGE_SIZE) /* finding contiguos pages can be hard*/
 		ret = um_kmalloc(size);
 	else ret = um_vmalloc(size);
 
_
