

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/video/modedb.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletion(-)

diff -puN drivers/video/modedb.c~video-mode-handling-linked-list-of-video-modes-build-fix drivers/video/modedb.c
--- 25/drivers/video/modedb.c~video-mode-handling-linked-list-of-video-modes-build-fix	2004-08-08 23:30:48.260406832 -0700
+++ 25-akpm/drivers/video/modedb.c	2004-08-08 23:31:03.227131544 -0700
@@ -683,9 +683,10 @@ struct fb_videomode *fb_find_best_mode(s
 	u32 diff = -1;
 
 	list_for_each(pos, head) {
+		u32 d;
+
 		modelist = list_entry(pos, struct fb_modelist, list);
 		mode = &modelist->mode;
-		u32 d;
 
 		if (mode->xres >= var->xres && mode->yres >= var->yres) {
 			d = (mode->xres - var->xres) +
_
