


---

 25-akpm/include/asm-x86_64/smp.h |    1 -
 1 files changed, 1 deletion(-)

diff -puN include/asm-x86_64/smp.h~x86_64-cpu_online-fix include/asm-x86_64/smp.h
--- 25/include/asm-x86_64/smp.h~x86_64-cpu_online-fix	2004-06-04 22:50:57.135811376 -0700
+++ 25-akpm/include/asm-x86_64/smp.h	2004-06-04 22:51:02.330021736 -0700
@@ -60,7 +60,6 @@ extern char phys_proc_id[NR_CPUS];
 
 extern cpumask_t cpu_callout_map;
 #define cpu_possible_map cpu_callout_map
-#define cpu_online(cpu) cpu_isset(cpu, cpu_online_map)
 
 static inline int num_booting_cpus(void)
 {

_
