

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/serial/Kconfig |    1 +
 1 files changed, 1 insertion(+)

diff -puN drivers/serial/Kconfig~altix-serial-driver-fix drivers/serial/Kconfig
--- 25/drivers/serial/Kconfig~altix-serial-driver-fix	Wed Jun 23 15:24:00 2004
+++ 25-akpm/drivers/serial/Kconfig	Wed Jun 23 15:24:41 2004
@@ -628,6 +628,7 @@ config SERIAL_LH7A40X_CONSOLE
 
 config SERIAL_SGI_L1_CONSOLE
 	bool "SGI Altix L1 serial console support"
+	depends on IA64_GENERIC || IA64_SGI_SN2
 	select SERIAL_CORE
 	help
 		If you have an SGI Altix and you would like to use the system
_
