
From: NeilBrown <neilb@cse.unsw.edu.au>

Signed-off-by: J. Bruce Fields <bfields@citi.umich.edu>         
Signed-off-by: Neil Brown <neilb@cse.unsw.edu.au>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/net/sunrpc/svcsock.c |    1 -
 1 files changed, 1 deletion(-)

diff -puN net/sunrpc/svcsock.c~knfsd-delete-an-obsolete-comment-from-nfsd-rpc-code net/sunrpc/svcsock.c
--- 25/net/sunrpc/svcsock.c~knfsd-delete-an-obsolete-comment-from-nfsd-rpc-code	2004-06-23 22:12:06.298709464 -0700
+++ 25-akpm/net/sunrpc/svcsock.c	2004-06-23 22:12:06.304708552 -0700
@@ -414,7 +414,6 @@ svc_sendto(struct svc_rqst *rqstp, struc
 	}
 	/* send tail */
 	if (xdr->tail[0].iov_len) {
-		/* The tail *will* be in respages[0]; */
 		result = sock->ops->sendpage(sock, rqstp->rq_respages[rqstp->rq_restailpage], 
 					     ((unsigned long)xdr->tail[0].iov_base)& (PAGE_SIZE-1),
 					     xdr->tail[0].iov_len, 0);
_
