

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/sys.c |    4 ++++
 1 files changed, 4 insertions(+)

diff -puN kernel/sys.c~add-missing-linux-syscallsh-includes-fix kernel/sys.c
--- 25/kernel/sys.c~add-missing-linux-syscallsh-includes-fix	Wed Sep 22 15:50:53 2004
+++ 25-akpm/kernel/sys.c	Wed Sep 22 15:51:38 2004
@@ -23,7 +23,11 @@
 #include <linux/security.h>
 #include <linux/dcookies.h>
 #include <linux/suspend.h>
+
+/* Don't include this - it breaks ia64's cond_syscall() implementation */
+#if 0
 #include <linux/syscalls.h>
+#endif
 
 #include <asm/uaccess.h>
 #include <asm/io.h>
_
