

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-power4-akpm/arch/ppc64/kernel/time.c |    1 +
 1 files changed, 1 insertion(+)

diff -puN arch/ppc64/kernel/time.c~bsd-secure-levels-lsm-add-time-hooks-ppc64-fix arch/ppc64/kernel/time.c
--- 25-power4/arch/ppc64/kernel/time.c~bsd-secure-levels-lsm-add-time-hooks-ppc64-fix	2004-10-06 23:59:34.262852800 -0700
+++ 25-power4-akpm/arch/ppc64/kernel/time.c	2004-10-06 23:59:34.266852192 -0700
@@ -49,6 +49,7 @@
 #include <linux/init.h>
 #include <linux/profile.h>
 #include <linux/cpu.h>
+#include <linux/security.h>
 
 #include <asm/segment.h>
 #include <asm/io.h>
_
