
From: Jan-Benedict Glaw <jbglaw@lug-owl.de>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/input/mouse/Kconfig |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff -puN drivers/input/mouse/Kconfig~document-dec-vsxxx-ab-digitizer-as-known-working drivers/input/mouse/Kconfig
--- 25/drivers/input/mouse/Kconfig~document-dec-vsxxx-ab-digitizer-as-known-working	Thu Sep 30 15:47:18 2004
+++ 25-akpm/drivers/input/mouse/Kconfig	Thu Sep 30 15:47:18 2004
@@ -126,6 +126,6 @@ config MOUSE_VSXXXAA
 	  puck) or a VSXXX-GA (rectangular) mouse. Theses mice are
 	  typically used on DECstations or VAXstations, but can also
 	  be used on any box capable of RS232 (with some adaptor
-	  described in the source file). This driver should, in theory,
-	  also work with the digitizer DEC produced, but it isn't tested
-	  with that (I don't have the hardware yet).
+	  described in the source file). This driver also works with the
+	  digitizer (VSXXX-AB) DEC produced.
+
_
