
From: Stephen Tweedie <sct@redhat.com>

Signed-off-by: Stephen Tweedie <sct@redhat.com>

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/include/linux/ext3_fs.h |    1 -
 1 files changed, 1 deletion(-)

diff -puN include/linux/ext3_fs.h~ext3-reservations-remove-unneeded-declaration include/linux/ext3_fs.h
--- 25/include/linux/ext3_fs.h~ext3-reservations-remove-unneeded-declaration	Tue Sep  7 15:36:16 2004
+++ 25-akpm/include/linux/ext3_fs.h	Tue Sep  7 15:36:16 2004
@@ -730,7 +730,6 @@ extern int  ext3_setattr (struct dentry 
 extern void ext3_put_inode (struct inode *);
 extern void ext3_delete_inode (struct inode *);
 extern int  ext3_sync_inode (handle_t *, struct inode *);
-extern void ext3_discard_prealloc (struct inode *);
 extern void ext3_discard_reservation (struct inode *);
 extern void ext3_dirty_inode(struct inode *);
 extern int ext3_change_inode_journal_flag(struct inode *, int);
_
