
From: OGAWA Hirofumi <hirofumi@mail.parknet.co.jp>

Signed-off-by: OGAWA Hirofumi <hirofumi@mail.parknet.co.jp>
---

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/fat/inode.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff -puN fs/fat/inode.c~fat-merge-fix fs/fat/inode.c
--- 25/fs/fat/inode.c~fat-merge-fix	2004-09-25 22:05:58.602708064 -0700
+++ 25-akpm/fs/fat/inode.c	2004-09-25 22:05:58.607707304 -0700
@@ -32,7 +32,7 @@ static int fat_default_codepage = CONFIG
 static char fat_default_iocharset[] = CONFIG_FAT_DEFAULT_IOCHARSET;
 
 static int fat_statfs(struct super_block *sb, struct kstatfs *buf);
-static void fat_write_inode(struct inode *inode, int wait);
+static int fat_write_inode(struct inode *inode, int wait);
 
 /*
  * New FAT inode stuff. We do the following:
@@ -1233,7 +1233,7 @@ static int fat_fill_inode(struct inode *
 	return 0;
 }
 
-int fat_write_inode(struct inode *inode, int wait)
+static int fat_write_inode(struct inode *inode, int wait)
 {
 	struct super_block *sb = inode->i_sb;
 	struct msdos_sb_info *sbi = MSDOS_SB(sb);
_
