
From: Nishanth Aravamudan <nacc@us.ibm.com>

Signed-off-by: Nishanth Aravamudan <nacc@us.ibm.com>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/net/mac89x0.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff -puN drivers/net/mac89x0.c~janitor-net-mac89x0-replace-schedule_timeout-with-msleep_interruptible drivers/net/mac89x0.c
--- 25/drivers/net/mac89x0.c~janitor-net-mac89x0-replace-schedule_timeout-with-msleep_interruptible	Thu Sep 23 17:18:30 2004
+++ 25-akpm/drivers/net/mac89x0.c	Thu Sep 23 17:18:30 2004
@@ -98,6 +98,7 @@ static char *version =
 #include <linux/netdevice.h>
 #include <linux/etherdevice.h>
 #include <linux/skbuff.h>
+#include <linux/delay.h>
 
 #include <asm/system.h>
 #include <asm/bitops.h>
@@ -308,8 +309,7 @@ void __init reset_chip(struct net_device
 	writereg(dev, PP_SelfCTL, readreg(dev, PP_SelfCTL) | POWER_ON_RESET);
 
 	/* wait 30 ms */
-	current->state = TASK_INTERRUPTIBLE;
-	schedule_timeout(30*HZ/1000);
+	msleep_interruptible(30);
 
 	/* Wait until the chip is reset */
 	reset_start_time = jiffies;
_
