
From: <janitor@sternwelten.at>

Signed-off-by: Maximilian Attems <janitor@sternwelten.at>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/net/bsd_comp.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff -puN drivers/net/bsd_comp.c~janitor-mark-__init-__exit-static-drivers-net-bsd_comp drivers/net/bsd_comp.c
--- 25/drivers/net/bsd_comp.c~janitor-mark-__init-__exit-static-drivers-net-bsd_comp	Thu Sep 23 15:45:38 2004
+++ 25-akpm/drivers/net/bsd_comp.c	Thu Sep 23 15:45:38 2004
@@ -1160,7 +1160,7 @@ static struct compressor ppp_bsd_compres
  * Module support routines
  *************************************************************/
 
-int __init bsdcomp_init(void)
+static int __init bsdcomp_init(void)
 {
 	int answer = ppp_register_compressor(&ppp_bsd_compress);
 	if (answer == 0)
@@ -1168,7 +1168,7 @@ int __init bsdcomp_init(void)
 	return answer;
 }
 
-void __exit bsdcomp_cleanup(void)
+static void __exit bsdcomp_cleanup(void)
 {
 	ppp_unregister_compressor(&ppp_bsd_compress);
 }
_
