
From: <janitor@sternwelten.at>

Signed-off-by: Maximilian Attems <janitor@sternwelten.at>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/net/ppp_deflate.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff -puN drivers/net/ppp_deflate.c~janitor-mark-__init-__exit-static-drivers-net-ppp_deflate drivers/net/ppp_deflate.c
--- 25/drivers/net/ppp_deflate.c~janitor-mark-__init-__exit-static-drivers-net-ppp_deflate	Thu Sep 23 15:45:16 2004
+++ 25-akpm/drivers/net/ppp_deflate.c	Thu Sep 23 15:45:16 2004
@@ -636,7 +636,7 @@ struct compressor ppp_deflate_draft = {
 	.owner =		THIS_MODULE
 };
 
-int __init deflate_init(void)
+static int __init deflate_init(void)
 {  
         int answer = ppp_register_compressor(&ppp_deflate);
         if (answer == 0)
@@ -646,7 +646,7 @@ int __init deflate_init(void)
         return answer;
 }
      
-void __exit deflate_cleanup(void)
+static void __exit deflate_cleanup(void)
 {
 	ppp_unregister_compressor(&ppp_deflate);
 	ppp_unregister_compressor(&ppp_deflate_draft);
_
