
From: <janitor@sternwelten.at>

Signed-off-by: Nishanth Aravamudan <nacc@us.ibm.com>
Signed-off-by: Maximilian Attems <janitor@sternwelten.at>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/char/sx.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN drivers/char/sx.c~janitor-char-sx-replace-direct-assignment-with-set_current_state drivers/char/sx.c
--- 25/drivers/char/sx.c~janitor-char-sx-replace-direct-assignment-with-set_current_state	2004-10-05 02:10:59.008738256 -0700
+++ 25-akpm/drivers/char/sx.c	2004-10-05 02:10:59.013737496 -0700
@@ -1499,7 +1499,7 @@ static void sx_close (void *ptr)
 	sx_send_command (port, HS_CLOSE, 0, 0);
 
 	while (to-- && (sx_read_channel_byte (port, hi_hstat) != HS_IDLE_CLOSED)) {
-		current->state = TASK_INTERRUPTIBLE;
+		set_current_state(TASK_INTERRUPTIBLE);
 		schedule_timeout (1);
 		if (signal_pending (current))
 				break;
_
