
From: "Sergey S. Kostyliov" <rathamahata@ehouse.ru>



Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/scsi/megaraid/megaraid_mbox.c |    6 ++++++
 1 files changed, 6 insertions(+)

diff -puN drivers/scsi/megaraid/megaraid_mbox.c~megaraid-random-loss-of-luns drivers/scsi/megaraid/megaraid_mbox.c
--- 25/drivers/scsi/megaraid/megaraid_mbox.c~megaraid-random-loss-of-luns	Fri Oct  8 15:12:09 2004
+++ 25-akpm/drivers/scsi/megaraid/megaraid_mbox.c	Fri Oct  8 15:12:09 2004
@@ -295,6 +295,12 @@ static struct pci_device_id pci_id_table
 		PCI_SUBSYS_ID_PERC3_SC,
 	},
 	{
+		PCI_VENDOR_ID_AMI,
+		PCI_DEVICE_ID_AMI_MEGARAID3,
+		PCI_VENDOR_ID_AMI,
+		PCI_SUBSYS_ID_PERC3_SC,
+	},
+	{
 		PCI_VENDOR_ID_LSI_LOGIC,
 		PCI_DEVICE_ID_MEGARAID_SCSI_320_0,
 		PCI_VENDOR_ID_LSI_LOGIC,
_
