Index: xx-sources/arch/i386/Kconfig
===================================================================
--- xx-sources.orig/arch/i386/Kconfig	2004-09-28 21:57:01.357362400 -0400
+++ xx-sources/arch/i386/Kconfig	2004-09-28 21:57:03.477040160 -0400
@@ -1285,3 +1285,5 @@
 	bool
 	depends on X86 && !EMBEDDED
 	default y
+
+source "kernel/Kconfig-extra.xx"
Index: xx-sources/kernel/Kconfig-extra.xx
===================================================================
--- xx-sources.orig/kernel/Kconfig-extra.xx	2004-05-31 17:36:38.000000000 -0400
+++ xx-sources/kernel/Kconfig-extra.xx	2004-09-28 21:58:07.238346968 -0400
@@ -0,0 +1,5 @@
+menu "Experimental xx options"
+
+source "kernel/sched/Kconfig"
+
+endmenu
Index: xx-sources/kernel/sched/Kconfig
===================================================================
--- xx-sources.orig/kernel/sched/Kconfig	2004-05-31 17:36:38.000000000 -0400
+++ xx-sources/kernel/sched/Kconfig	2004-09-28 21:57:48.180244240 -0400
@@ -0,0 +1,18 @@
+choice
+	prompt "Process scheduling policy"
+	default SCHED_NONE
+	depends on EXPERIMENTAL
+	help
+	  Different people have written alternate implementations of
+	  the kernel's scheduler code.  Each implementation is different
+	  and has its own pros and cons.  This will allow you to choose
+	  between different scheduling policies.
+
+config SCHED_NONE
+	bool "Default"
+	help
+	  This is the default scheduler as is included in the -mm kernels.
+	  It contains the sched domains code by Nick Piggin and some tweaks
+	  to the scheduling code, but no significant changes.
+
+endchoice
