Index: xx-sources/init/Kconfig
===================================================================
--- xx-sources.orig/init/Kconfig	2004-09-13 17:14:14.979135160 -0400
+++ xx-sources/init/Kconfig	2004-09-13 17:15:00.436224632 -0400
@@ -308,19 +308,6 @@
 
 	  Say N if unsure.
 
-config CC_OPTIMIZE_FOR_SIZE
-	bool "Optimize for size" if EMBEDDED
-	default y if ARM || H8300
-	default n
-	help
-	  Enabling this option will pass "-Os" instead of "-O2" to gcc
-	  resulting in a smaller kernel.
-
-	  WARNING: some versions of gcc may generate incorrect code with this
-	  option.  If problems are observed, a gcc upgrade may be needed.
-
-	  If unsure, say N.
-
 config SHMEM
 	default y
 	bool "Use full shmem filesystem" if EMBEDDED && MMU
Index: xx-sources/kernel/Kconfig-extra.xx
===================================================================
--- xx-sources.orig/kernel/Kconfig-extra.xx	2004-09-13 17:14:33.096380920 -0400
+++ xx-sources/kernel/Kconfig-extra.xx	2004-09-13 17:14:42.054019152 -0400
@@ -167,6 +167,19 @@
 	  no dummy operations need be executed.
 	  Zero means use compiler's default.
 
+config CC_OPTIMIZE_FOR_SIZE
+	bool "Optimize for size"
+	default y if ARM || H8300
+	default n
+	help
+	  Enabling this option will pass "-Os" instead of "-O2" to gcc
+	  resulting in a smaller kernel.
+
+	  WARNING: some versions of gcc may generate incorrect code with this
+	  option.  If problems are observed, a gcc upgrade may be needed.
+
+	  If unsure, say N.
+
 endmenu
 
 config HERTZ
