

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/inode.c |    1 -
 1 files changed, 1 deletion(-)

diff -puN fs/inode.c~reiser4-export-inode_lock-cleanup fs/inode.c
--- 25/fs/inode.c~reiser4-export-inode_lock-cleanup	2004-10-03 16:43:06.019416464 -0700
+++ 25-akpm/fs/inode.c	2004-10-03 16:43:06.022416008 -0700
@@ -233,7 +233,6 @@ void __iget(struct inode * inode)
 		list_move(&inode->i_list, &inode_in_use);
 	inodes_stat.nr_unused--;
 }
-
 EXPORT_SYMBOL(__iget);
 
 /**
_
