

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/inode.c |    4 ----
 1 files changed, 4 deletions(-)

diff -puN fs/inode.c~reiser4-allow-drop_inode-implementation-cleanup fs/inode.c
--- 25/fs/inode.c~reiser4-allow-drop_inode-implementation-cleanup	2004-10-03 16:43:04.441656320 -0700
+++ 25-akpm/fs/inode.c	2004-10-03 16:43:04.445655712 -0700
@@ -181,10 +181,8 @@ void destroy_inode(struct inode *inode) 
 	else
 		kmem_cache_free(inode_cachep, (inode));
 }
-
 EXPORT_SYMBOL(destroy_inode);
 
-
 /*
  * These are initializations that only need to be done
  * once, because the fields are idempotent across use
@@ -1061,7 +1059,6 @@ void generic_forget_inode(struct inode *
 	clear_inode(inode);
 	destroy_inode(inode);
 }
-
 EXPORT_SYMBOL(generic_forget_inode);
 
 /*
@@ -1335,7 +1332,6 @@ void wake_up_inode(struct inode *inode)
 	if (waitqueue_active(wq))
 		wake_up_all(wq);
 }
-
 EXPORT_SYMBOL(wake_up_inode);
 
 static __initdata unsigned long ihash_entries;
_
