

insanity.

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/reiser4/debug.c |    2 --
 1 files changed, 2 deletions(-)

diff -puN fs/reiser4/debug.c~reiser4-cond_resched-build-fix fs/reiser4/debug.c
--- 25/fs/reiser4/debug.c~reiser4-cond_resched-build-fix	2004-09-15 22:10:17.042384152 -0700
+++ 25-akpm/fs/reiser4/debug.c	2004-09-15 22:10:17.047383392 -0700
@@ -45,8 +45,6 @@
 #include <linux/ctype.h>
 #include <linux/sysctl.h>
 
-extern void cond_resched(void);
-
 /*
  * global buffer where message given to reiser4_panic is formatted.
  */
_
