

hrm.

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-power4-akpm/kernel/sched.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff -puN kernel/sched.c~sched-fix-sched_smt-numa=fake=2-lockup-fix kernel/sched.c
--- 25-power4/kernel/sched.c~sched-fix-sched_smt-numa=fake=2-lockup-fix	2004-10-10 00:57:20.172717024 -0700
+++ 25-power4-akpm/kernel/sched.c	2004-10-10 00:57:44.696988768 -0700
@@ -4520,8 +4520,10 @@ static void check_sibling_maps(void)
 
 	for_each_online_cpu(i) {
 		for_each_cpu_mask(j, cpu_sibling_map[i]) {
-			if (cpu_to_node[i] != cpu_to_node[j]) {
-				printk(KERN_INFO "warning: CPU %d siblings map to different node - isolating them.\n", i);
+			if (cpu_to_node(i) != cpu_to_node(j)) {
+				printk(KERN_INFO "warning: CPU %d siblings map "
+					"to different node - isolating "
+					"them.\n", i);
 				cpu_sibling_map[i] = cpumask_of_cpu(i);
 				break;
 			}
_
