

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/sched.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletion(-)

diff -puN kernel/sched.c~sched-print-preempt-count kernel/sched.c
--- 25/kernel/sched.c~sched-print-preempt-count	2004-10-05 02:09:50.226194792 -0700
+++ 25-akpm/kernel/sched.c	2004-10-05 02:09:50.232193880 -0700
@@ -2492,7 +2492,9 @@ asmlinkage void __sched schedule(void)
 	 */
 	if (likely(!(current->state & (TASK_DEAD | TASK_ZOMBIE)))) {
 		if (unlikely(in_atomic())) {
-			printk(KERN_ERR "bad: scheduling while atomic!\n");
+			printk(KERN_ERR "scheduling while atomic: "
+				"%s/0x%08x/%d\n",
+				current->comm, preempt_count(), current->pid);
 			dump_stack();
 		}
 	}
_
