
From: Christoph Hellwig <hch@lst.de>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/fcntl.c |    2 --
 1 files changed, 2 deletions(-)

diff -puN fs/fcntl.c~unexport-f_delown fs/fcntl.c
--- 25/fs/fcntl.c~unexport-f_delown	2004-09-21 01:57:21.992094512 -0700
+++ 25-akpm/fs/fcntl.c	2004-09-21 01:57:21.995094056 -0700
@@ -291,8 +291,6 @@ void f_delown(struct file *filp)
 	f_modown(filp, 0, 0, 0, 1);
 }
 
-EXPORT_SYMBOL(f_delown);
-
 static long do_fcntl(int fd, unsigned int cmd, unsigned long arg,
 		struct file *filp)
 {
_
