

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/perfctr/virtual.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff -puN drivers/perfctr/virtual.c~perfctr-cpus_complement-fix drivers/perfctr/virtual.c
--- 25/drivers/perfctr/virtual.c~perfctr-cpus_complement-fix	2004-06-06 23:40:06.774412408 -0700
+++ 25-akpm/drivers/perfctr/virtual.c	2004-06-06 23:40:37.435751176 -0700
@@ -403,10 +403,11 @@ static int do_vperfctr_control(struct vp
 		return -EFAULT;
 
 	if (control.cpu_control.nractrs || control.cpu_control.nrictrs) {
-		cpumask_t tmp, old_mask, new_mask;
+		cpumask_t tmp, old_mask, new_mask, tmp1;
 
 		tmp = perfctr_cpus_forbidden_mask;
-		cpus_complement(tmp);
+		cpus_complement(tmp1, tmp);
+		tmp = tmp1;
 		old_mask = tsk->cpus_allowed;
 		cpus_and(new_mask, old_mask, tmp);
 
_
