
From: Eberhard Moenkeberg <emoenke@gwdg.de>



Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/CREDITS |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff -puN CREDITS~credits-update CREDITS
--- 25/CREDITS~credits-update	2004-06-19 14:17:41.195248904 -0700
+++ 25-akpm/CREDITS	2004-06-19 14:17:41.200248144 -0700
@@ -2267,8 +2267,8 @@ S: USA
 N: Eberhard Moenkeberg
 E: emoenke@gwdg.de
 D: CDROM driver "sbpcd" (Matsushita/Panasonic/Soundblaster)
-S: Reinholdstrasse 14
-S: D-37083 Goettingen
+S: Ruhstrathoehe 2 b.
+S: D-37085 Goettingen
 S: Germany
 
 N: Thomas Molina
_
