
From: FabF <fabian.frederick@skynet.be>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/quota.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN fs/quota.c~sparsify-quotactl fs/quota.c
--- 25/fs/quota.c~sparsify-quotactl	2004-06-19 14:06:37.144200008 -0700
+++ 25-akpm/fs/quota.c	2004-06-19 14:06:37.148199400 -0700
@@ -264,7 +264,7 @@ static int do_quotactl(struct super_bloc
  * calls. Maybe we need to add the process quotas etc. in the future,
  * but we probably should use rlimits for that.
  */
-asmlinkage long sys_quotactl(unsigned int cmd, const char *special, qid_t id, caddr_t addr)
+asmlinkage long sys_quotactl(unsigned int cmd, const char __user *special, qid_t id, caddr_t addr)
 {
 	uint cmds, type;
 	struct super_block *sb = NULL;
_
