

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/drivers/input/serio/i8042.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletion(-)

diff -puN drivers/input/serio/i8042.c~input-serio-dynamic-allocation-fix drivers/input/serio/i8042.c
--- 25/drivers/input/serio/i8042.c~input-serio-dynamic-allocation-fix	2004-06-19 22:53:31.959017616 -0700
+++ 25-akpm/drivers/input/serio/i8042.c	2004-06-19 22:53:31.964016856 -0700
@@ -25,7 +25,6 @@
 #include <asm/io.h>
 
 #undef DEBUG
-#include "i8042.h"
 
 MODULE_AUTHOR("Vojtech Pavlik <vojtech@suse.cz>");
 MODULE_DESCRIPTION("i8042 keyboard and mouse controller driver");
@@ -55,6 +54,8 @@ static unsigned int i8042_dumbkbd;
 module_param_named(dumbkbd, i8042_dumbkbd, bool, 0);
 MODULE_PARM_DESC(dumbkbd, "Pretend that controller can only read data from keyboard");
 
+#include "i8042.h"
+
 extern unsigned int i8042_dmi_noloop;
 static unsigned int i8042_noloop;
 extern unsigned int i8042_dmi_noloop;
_
