
From: Dave Kleikamp <shaggy@austin.ibm.com>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/jfs/jfs_dtree.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletion(-)

diff -puN fs/jfs/jfs_dtree.c~jfs-build-fix fs/jfs/jfs_dtree.c
--- 25/fs/jfs/jfs_dtree.c~jfs-build-fix	Wed Jun 16 15:09:33 2004
+++ 25-akpm/fs/jfs/jfs_dtree.c	Wed Jun 16 15:09:33 2004
@@ -374,6 +374,8 @@ static u32 add_index(tid_t tid, struct i
 		return index;
 	}
 	if (index == (MAX_INLINE_DIRTABLE_ENTRY + 1)) {
+		struct dir_table_slot temp_table[12];
+
 		/*
 		 * It's time to move the inline table to an external
 		 * page and begin to build the xtree
@@ -385,7 +387,6 @@ static u32 add_index(tid_t tid, struct i
 		 * Save the table, we're going to overwrite it with the
 		 * xtree root
 		 */
-		struct dir_table_slot temp_table[12];
 		memcpy(temp_table, &jfs_ip->i_dirtable, sizeof(temp_table));
 
 		/*
_
