---

 linux-2.6.7-rc2-xx3-xiphux/kernel/sched.c |    9 +++++++++
 1 files changed, 9 insertions(+)

diff -puN fs/select.c~21-software-suspend-linux-2.6.6-test10-incremental fs/select.c
diff -puN kernel/sched.c~21-software-suspend-linux-2.6.6-test10-incremental kernel/sched.c
--- linux-2.6.7-rc2-xx3/kernel/sched.c~21-software-suspend-linux-2.6.6-test10-incremental	2004-06-04 00:41:51.000000000 -0400
+++ linux-2.6.7-rc2-xx3-xiphux/kernel/sched.c	2004-06-04 01:41:52.403434208 -0400
@@ -3126,6 +3126,15 @@ need_resched:
 	}
 #endif
 	/*
+	 * if this task is only one running, don't preempt it
+	 * with idle thread.
+	 */
+
+	if (unlikely((preempt_count() & PREEMPT_ACTIVE) &&
+		(rq->nr_running == 1)))
+		return;
+
+	/*
 	 * if entering off of a kernel preemption go straight
 	 * to picking the next task.
 	 */

_
