diff -uNr linux-2.6.8.1-mm4/include/linux/fs.h linux-2.6.8.1-mm4-cachefs/include/linux/fs.h
--- linux-2.6.8.1-mm4/include/linux/fs.h	2004-08-23 13:09:06.000000000 +0100
+++ linux-2.6.8.1-mm4-cachefs/include/linux/fs.h	2004-08-25 19:39:35.000000000 +0100
@@ -1076,7 +1076,7 @@
 
 static inline void file_accessed(struct file *file)
 {
-	if (!(file->f_flags & O_NOATIME))
+	if (file && !(file->f_flags & O_NOATIME))
 		touch_atime(file->f_vfsmnt, file->f_dentry);
 }
 
