diff -uNr linux-2.6.8.1-mm4/mm/readahead.c linux-2.6.8.1-mm4-cachefs/mm/readahead.c
--- linux-2.6.8.1-mm4/mm/readahead.c	2004-08-23 13:09:09.000000000 +0100
+++ linux-2.6.8.1-mm4-cachefs/mm/readahead.c	2004-08-23 14:33:07.000000000 +0100
@@ -34,9 +34,11 @@
 void
 file_ra_state_init(struct file_ra_state *ra, struct address_space *mapping)
 {
+	memset(ra, 0, sizeof(*ra));
 	ra->ra_pages = mapping->backing_dev_info->ra_pages;
 	ra->average = ra->ra_pages / 2;
 }
+EXPORT_SYMBOL_GPL(file_ra_state_init);
 
 /*
  * Return max readahead size for this inode in number-of-pages.
