

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/filemap.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff -puN mm/filemap.c~consolidate-bit-waiting-code-patterns-cleanup mm/filemap.c
--- 25/mm/filemap.c~consolidate-bit-waiting-code-patterns-cleanup	2004-08-28 15:31:51.888527664 -0700
+++ 25-akpm/mm/filemap.c	2004-08-28 15:32:00.237258464 -0700
@@ -135,8 +135,9 @@ EXPORT_SYMBOL(remove_from_page_cache);
 static int sync_page(void *word)
 {
 	struct address_space *mapping;
-	struct page *page
-		= container_of((page_flags_t *)word, struct page, flags);
+	struct page *page;
+
+	page = container_of((page_flags_t *)word, struct page, flags);
 
 	/*
 	 * FIXME, fercrissake.  What is this barrier here for?
_
