
From: Joshua Kwan <joshk@triplehelix.org>

Cc: William Lee Irwin III <wli@holomorphy.com>

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/sparc/kernel/sparc_ksyms.c |    7 +++++++
 1 files changed, 7 insertions(+)

diff -puN arch/sparc/kernel/sparc_ksyms.c~export-more-symbols-on-sparc32 arch/sparc/kernel/sparc_ksyms.c
--- 25/arch/sparc/kernel/sparc_ksyms.c~export-more-symbols-on-sparc32	2004-08-28 14:53:40.985797816 -0700
+++ 25-akpm/arch/sparc/kernel/sparc_ksyms.c	2004-08-28 14:53:40.988797360 -0700
@@ -214,6 +214,10 @@ EXPORT_SYMBOL(pci_dma_sync_single_for_cp
 EXPORT_SYMBOL(pci_dma_sync_single_for_device);
 EXPORT_SYMBOL(pci_dma_sync_sg_for_cpu);
 EXPORT_SYMBOL(pci_dma_sync_sg_for_device);
+EXPORT_SYMBOL(pci_map_sg);
+EXPORT_SYMBOL(pci_unmap_sg);
+EXPORT_SYMBOL(pci_map_page);
+EXPORT_SYMBOL(pci_unmap_page);
 /* Actually, ioremap/iounmap are not PCI specific. But it is ok for drivers. */
 EXPORT_SYMBOL(ioremap);
 EXPORT_SYMBOL(iounmap);
@@ -298,6 +302,9 @@ EXPORT_SYMBOL(csum_partial);
 /* Cache flushing.  */
 EXPORT_SYMBOL(sparc_flush_page_to_ram);
 
+/* For when serial stuff is built as modules. */
+EXPORT_SYMBOL(sun_do_break);
+
 EXPORT_SYMBOL(__ret_efault);
 
 EXPORT_SYMBOL(memcmp);
_
