
From: Zwane Mwaikambo <zwane@linuxpower.ca>

Signed-off-by: Zwane Mwaikambo <zwane@linuxpower.ca>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/i386/kernel/io_apic.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN arch/i386/kernel/io_apic.c~find_isa_irq_pin-should-not-be-__init arch/i386/kernel/io_apic.c
--- 25/arch/i386/kernel/io_apic.c~find_isa_irq_pin-should-not-be-__init	Mon Aug 23 15:59:20 2004
+++ 25-akpm/arch/i386/kernel/io_apic.c	Mon Aug 23 15:59:20 2004
@@ -747,7 +747,7 @@ static int __init find_irq_entry(int api
 /*
  * Find the pin to which IRQ[irq] (ISA) is connected
  */
-static int __init find_isa_irq_pin(int irq, int type)
+static int find_isa_irq_pin(int irq, int type)
 {
 	int i;
 
_
