
From: "Luiz Fernando N. Capitulino" <lcapitulino@conectiva.com.br>

Signed-off-by: Luiz Capitulino <lcapitulino@conectiva.com.br>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/fork.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN kernel/fork.c~kernel-forkc-add-missing-unlikely kernel/fork.c
--- 25/kernel/fork.c~kernel-forkc-add-missing-unlikely	Mon Aug 30 14:42:19 2004
+++ 25-akpm/kernel/fork.c	Mon Aug 30 14:42:19 2004
@@ -1005,7 +1005,7 @@ static task_t *copy_process(unsigned lon
 	}
 
 	SET_LINKS(p);
-	if (p->ptrace & PT_PTRACED)
+	if (unlikely(p->ptrace & PT_PTRACED))
 		__ptrace_link(p, current->parent);
 
 	cpuset_fork(p);
_
