

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/i386/Kconfig.kgdb |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN arch/i386/Kconfig.kgdb~kgdb-is-incompatible-with-kprobes arch/i386/Kconfig.kgdb
--- 25/arch/i386/Kconfig.kgdb~kgdb-is-incompatible-with-kprobes	2004-08-16 12:19:40.461128960 -0700
+++ 25-akpm/arch/i386/Kconfig.kgdb	2004-08-16 12:19:51.264486600 -0700
@@ -1,6 +1,6 @@
 config KGDB
 	bool "Include kgdb kernel debugger"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !KPROBES
 	help
 	  If you say Y here, the system will be compiled with the debug
 	  option (-g) and a debugging stub will be included in the
_
