

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/lockmeter.c |    1 +
 1 files changed, 1 insertion(+)

diff -puN kernel/lockmeter.c~lockmeter-build-fix kernel/lockmeter.c
--- 25/kernel/lockmeter.c~lockmeter-build-fix	2004-08-18 22:35:45.501995600 -0700
+++ 25-akpm/kernel/lockmeter.c	2004-08-18 22:36:00.289747520 -0700
@@ -1175,4 +1175,5 @@ EXPORT_SYMBOL(_metered_spin_trylock);
 EXPORT_SYMBOL(_metered_read_lock);
 EXPORT_SYMBOL(_metered_read_unlock);
 EXPORT_SYMBOL(_metered_write_lock);
+EXPORT_SYMBOL(_metered_write_trylock);
 EXPORT_SYMBOL(_metered_write_unlock);
_
