
Make gconfig and menuconfig display the kernel name.


Index: xx-elv-select-cfqionice/Makefile
===================================================================
--- xx-elv-select-cfqionice.orig/Makefile	2004-08-19 05:29:14.000000000 -0400
+++ xx-elv-select-cfqionice/Makefile	2004-08-19 22:50:16.829793216 -0400
@@ -309,7 +309,7 @@
 export	VERSION PATCHLEVEL SUBLEVEL EXTRAVERSION KERNELRELEASE ARCH \
 	CONFIG_SHELL HOSTCC HOSTCFLAGS CROSS_COMPILE AS LD CC \
 	CPP AR NM STRIP OBJCOPY OBJDUMP MAKE AWK GENKSYMS PERL UTS_MACHINE \
-	HOSTCXX HOSTCXXFLAGS LDFLAGS_BLOB LDFLAGS_MODULE CHECK CHECKFLAGS
+	HOSTCXX HOSTCXXFLAGS LDFLAGS_BLOB LDFLAGS_MODULE CHECK CHECKFLAGS NAME
 
 export CPPFLAGS NOSTDINC_FLAGS LINUXINCLUDE OBJCOPYFLAGS LDFLAGS
 export CFLAGS CFLAGS_KERNEL CFLAGS_MODULE 
Index: xx-elv-select-cfqionice/scripts/kconfig/gconf.c
===================================================================
--- xx-elv-select-cfqionice.orig/scripts/kconfig/gconf.c	2004-08-19 05:03:34.000000000 -0400
+++ xx-elv-select-cfqionice/scripts/kconfig/gconf.c	2004-08-19 22:49:56.291915448 -0400
@@ -275,9 +275,10 @@
 					  /*"style", PANGO_STYLE_OBLIQUE, */
 					  NULL);
 
-	sprintf(title, "Linux Kernel v%s.%s.%s%s Configuration",
+	sprintf(title, "Linux Kernel v%s.%s.%s%s \"%s\" Configuration",
 		getenv("VERSION"), getenv("PATCHLEVEL"),
-		getenv("SUBLEVEL"), getenv("EXTRAVERSION"));
+		getenv("SUBLEVEL"), getenv("EXTRAVERSION"),
+		getenv("NAME"));
 	gtk_window_set_title(GTK_WINDOW(main_wnd), title);
 
 	gtk_widget_show(main_wnd);
Index: xx-elv-select-cfqionice/scripts/kconfig/mconf.c
===================================================================
--- xx-elv-select-cfqionice.orig/scripts/kconfig/mconf.c	2004-08-19 05:33:33.000000000 -0400
+++ xx-elv-select-cfqionice/scripts/kconfig/mconf.c	2004-08-19 22:49:56.292915296 -0400
@@ -768,8 +768,8 @@
 
 	sym = sym_lookup("KERNELRELEASE", 0);
 	sym_calc_value(sym);
-	sprintf(menu_backtitle, "Linux Kernel v%s Configuration",
-		sym_get_string_value(sym));
+	sprintf(menu_backtitle, "Linux Kernel v%s \"%s\" Configuration",
+		sym_get_string_value(sym), getenv("NAME"));
 
 	mode = getenv("MENUCONFIG_MODE");
 	if (mode) {
