
From: Nick Piggin <nickpiggin@yahoo.com.au>



Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/include/linux/pid.h |    3 ++-
 1 files changed, 2 insertions(+), 1 deletion(-)

diff -puN include/linux/pid.h~use-hlist-for-pid-hash-cache-friendliness include/linux/pid.h
--- 25/include/linux/pid.h~use-hlist-for-pid-hash-cache-friendliness	2004-08-21 23:42:27.169260240 -0700
+++ 25-akpm/include/linux/pid.h	2004-08-21 23:42:27.173259632 -0700
@@ -12,11 +12,12 @@ enum pid_type
 
 struct pid
 {
+	/* Try to keep hash_chain in the same cacheline as nr for find_pid */
+	struct hlist_node hash_chain;
 	int nr;
 	atomic_t count;
 	struct task_struct *task;
 	struct list_head task_list;
-	struct hlist_node hash_chain;
 };
 
 struct pid_link
_
