diff -ruN software-suspend-core-2.0.0.76/include/linux/suspend-common.h software-suspend-core-2.0.0.77/include/linux/suspend-common.h
--- software-suspend-core-2.0.0.76/include/linux/suspend-common.h	2004-05-19 12:43:03.000000000 +1000
+++ software-suspend-core-2.0.0.77/include/linux/suspend-common.h	2004-05-19 12:43:07.000000000 +1000
@@ -360,7 +360,7 @@
  */
 #if defined(CONFIG_SOFTWARE_SUSPEND_DEBUG)
 #define MDELAY(a) do { if (TEST_ACTION_STATE(SUSPEND_SLOW)) mdelay(a); } \
-	while (0);
+	while (0)
 
 #else // #ifdef CONFIG_SOFTWARE_SUSPEND_DEBUG
 
diff -ruN software-suspend-core-2.0.0.76/include/linux/suspend-debug.h software-suspend-core-2.0.0.77/include/linux/suspend-debug.h
--- software-suspend-core-2.0.0.76/include/linux/suspend-debug.h	2004-05-19 12:43:03.000000000 +1000
+++ software-suspend-core-2.0.0.77/include/linux/suspend-debug.h	2004-05-19 12:43:07.000000000 +1000
@@ -5,7 +5,7 @@
 /* Solely for comparison with the version specific patch revision */
 #define SWSUSP_CORE_REVISION 0x201
 
-#define SWSUSP_CORE_VERSION "2.0.0.76"
+#define SWSUSP_CORE_VERSION "2.0.0.77"
 #define name_suspend "Software Suspend " SWSUSP_CORE_VERSION ": "
 
 /* Same length to ensure one overwrites the other */
@@ -112,8 +112,6 @@
 #define CHECKMASK(mask) (((!mask) || (TEST_DEBUG_STATE(mask))) \
 		&& (!NO_OUTPUT_OR_PAUSING))
 
-//void __printnolog(int mask, int level, int restartline, const char *fmt, ...);
-//void __printlog(int mask, int level, const char *fmt, ...);
 void __printnolog(int restartline, const char *fmt, ...);
 void __printlog(const char *fmt, ...);
 
diff -ruN software-suspend-core-2.0.0.76/kernel/power/process.c software-suspend-core-2.0.0.77/kernel/power/process.c
--- software-suspend-core-2.0.0.76/kernel/power/process.c	2004-05-19 12:43:03.000000000 +1000
+++ software-suspend-core-2.0.0.77/kernel/power/process.c	2004-05-19 12:43:07.000000000 +1000
@@ -210,7 +210,7 @@
 		const char * file)
 {
 	if (unlikely(((suspend_task) && (current->pid == suspend_task)) ||
-		     (current->flags & PF_NOFREEZE)));
+		     (current->flags & PF_NOFREEZE)))
 		return;
 
 	if (unlikely(current->flags & PF_FREEZE))
