---

 linux-2.6.7-rc2-xx3-xiphux/Makefile |   17 +++++++++++++++--
 1 files changed, 15 insertions(+), 2 deletions(-)

diff -puN Makefile~scheduler-naming Makefile
--- linux-2.6.7-rc2-xx3/Makefile~scheduler-naming	2004-06-07 21:18:05.407101144 -0400
+++ linux-2.6.7-rc2-xx3-xiphux/Makefile	2004-06-07 21:18:05.459093240 -0400
@@ -25,6 +25,19 @@ export INTEL_CC INTEL_INC INTEL_PGOPTI
 
 NAME=Zonked Quokka
 
+ifeq ($(CONFIG_NICKSCHED),1)
+SCHEDULER =-nick
+endif
+ifeq ($(CONFIG_STAIRCASE),1)
+SCHEDULER =-stair
+endif
+ifeq ($(CONFIG_SPA),1)
+SCHEDULER =-spa
+endif
+ifeq ($(CONFIG_EBS),1)
+SCHEDULER =-ebs
+endif
+
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
 # More info can be located in ./README
@@ -172,7 +185,7 @@ VPATH		:= $(srctree)
 
 export srctree objtree VPATH TOPDIR
 
-KERNELRELEASE=$(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
+KERNELRELEASE=$(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)$(SCHEDULER)
 
 # SUBARCH tells the usermode build what the underlying arch is.  That is set
 # first, and if a usermode build is happening, the "ARCH=um" on the command
@@ -354,7 +367,7 @@ endif
 
 AFLAGS		:= -D__ASSEMBLY__
 
-export	VERSION PATCHLEVEL SUBLEVEL EXTRAVERSION KERNELRELEASE ARCH \
+export	VERSION PATCHLEVEL SUBLEVEL EXTRAVERSION SCHEDULER KERNELRELEASE ARCH \
 	CONFIG_SHELL HOSTCC HOSTCFLAGS CROSS_COMPILE AS LD CC \
 	CPP AR NM STRIP OBJCOPY OBJDUMP MAKE AWK GENKSYMS PERL UTS_MACHINE \
 	HOSTCXX HOSTCXXFLAGS LDFLAGS_BLOB LDFLAGS_MODULE CHECK NAME

_
