diff -ruN software-suspend-core-2.0.0.72/include/linux/suspend-debug.h software-suspend-core-2.0.0.73/include/linux/suspend-debug.h
--- software-suspend-core-2.0.0.72/include/linux/suspend-debug.h	2004-05-13 16:48:38.000000000 +1000
+++ software-suspend-core-2.0.0.73/include/linux/suspend-debug.h	2004-05-13 16:48:47.000000000 +1000
@@ -5,7 +5,7 @@
 /* Solely for comparison with the version specific patch revision */
 #define SWSUSP_CORE_REVISION 0x201
 
-#define SWSUSP_CORE_VERSION "2.0.0.72"
+#define SWSUSP_CORE_VERSION "2.0.0.73"
 #define name_suspend "Software Suspend " SWSUSP_CORE_VERSION ": "
 
 /* Same length to ensure one overwrites the other */
diff -ruN software-suspend-core-2.0.0.72/kernel/power/swapwriter.c software-suspend-core-2.0.0.73/kernel/power/swapwriter.c
--- software-suspend-core-2.0.0.72/kernel/power/swapwriter.c	2004-05-13 16:48:38.000000000 +1000
+++ software-suspend-core-2.0.0.73/kernel/power/swapwriter.c	2004-05-13 16:48:47.000000000 +1000
@@ -26,7 +26,9 @@
 #include <linux/suspend-common.h>
 
 struct suspend_plugin_ops swapwriterops;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0)
 static void sync_swap_partitions(void);
+#endif
 
 #define SIGNATURE_VER 6
 
