---

 linux-2.6.5-xx21-xiphux/arch/i386/Kconfig |   88 +++++++++++++++---------------
 1 files changed, 44 insertions(+), 44 deletions(-)

diff -puN arch/i386/Kconfig~preempt-kconfig-relocate-xx arch/i386/Kconfig
--- linux-2.6.5-xx21/arch/i386/Kconfig~preempt-kconfig-relocate-xx	2004-05-09 14:05:20.095973080 -0400
+++ linux-2.6.5-xx21-xiphux/arch/i386/Kconfig	2004-05-09 14:05:52.140101632 -0400
@@ -603,50 +603,6 @@ config SCHED_SMT
 	  cost of slightly increased overhead in some places. If unsure say
 	  N here.
 
-choice
-	prompt "Latency management"
-	default LATENCY_NONE
-	help
-	  This will allow you to choose between different ways to improve
-	  the latency of the kernel.  These options are, for the moment,
-	  mutually exclusive.
-
-	  Nothing is selected by default.
-
-config LATENCY_NONE
-	bool "None"
-	help
-	  With this option, the kernel will not be enhanced to be preemptible
-	  nor low latency.  This is not recommended for a desktop system, and
-	  should really only be used for troubleshooting.
-
-config PREEMPT
-	bool "Preemptible Kernel"
-	help
-	  This option reduces the latency of the kernel when reacting to
-	  real-time or interactive events by allowing a low priority process to
-	  be preempted even if it is in kernel mode executing a system call.
-	  This allows applications to run more reliably even when the system is
-	  under load.
-
-	  Say Y here if you are building a kernel for a desktop, embedded
-	  or real-time system.  Say N if you are unsure.  Low-latency
-	  must be disabled.
-
-config LOW_LATENCY
-	bool "Low-latency Kernel"
-	depends on EXPERIMENTAL
-	help
-	  Activating the low-latency patch theoretically fixes 90% of
-	  latency issues and supposedly outperforms preempt.  This
-	  config option is only temporary, for testing.
-
-	  Say Y if you want to test the low latency calls instead of the
-	  standard preempt.  Preempt must be disabled.  Say N if you
-	  are unsure.
-
-endchoice
-
 config X86_UP_APIC
 	bool "Local APIC support on uniprocessors" if !SMP
 	depends on !(X86_VISWS || X86_VOYAGER)
@@ -1718,4 +1674,48 @@ config VM_AUTOSWAPPINESS
 	  this does is sets whether the setting defaults to on or off
 	  when the machine boots.
 
+choice
+	prompt "Latency management"
+	default LATENCY_NONE
+	help
+	  This will allow you to choose between different ways to improve
+	  the latency of the kernel.  These options are, for the moment,
+	  mutually exclusive.
+
+	  Nothing is selected by default.
+
+config LATENCY_NONE
+	bool "None"
+	help
+	  With this option, the kernel will not be enhanced to be preemptible
+	  nor low latency.  This is not recommended for a desktop system, and
+	  should really only be used for troubleshooting.
+
+config PREEMPT
+	bool "Preemptible Kernel"
+	help
+	  This option reduces the latency of the kernel when reacting to
+	  real-time or interactive events by allowing a low priority process to
+	  be preempted even if it is in kernel mode executing a system call.
+	  This allows applications to run more reliably even when the system is
+	  under load.
+
+	  Say Y here if you are building a kernel for a desktop, embedded
+	  or real-time system.  Say N if you are unsure.  Low-latency
+	  must be disabled.
+
+config LOW_LATENCY
+	bool "Low-latency Kernel"
+	depends on EXPERIMENTAL
+	help
+	  Activating the low-latency patch theoretically fixes 90% of
+	  latency issues and supposedly outperforms preempt.  This
+	  config option is only temporary, for testing.
+
+	  Say Y if you want to test the low latency calls instead of the
+	  standard preempt.  Preempt must be disabled.  Say N if you
+	  are unsure.
+
+endchoice
+
 endmenu

_
