

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/afs/file.c |    4 ++++
 1 files changed, 4 insertions(+)

diff -puN fs/afs/file.c~afs-cachefs-dependency-fix fs/afs/file.c
--- 25/fs/afs/file.c~afs-cachefs-dependency-fix	2004-09-20 10:56:28.714441752 -0700
+++ 25-akpm/fs/afs/file.c	2004-09-20 10:57:39.770639568 -0700
@@ -33,7 +33,9 @@ static int afs_file_releasepage(struct p
 
 static ssize_t afs_file_write(struct file *file, const char __user *buf,
 			      size_t size, loff_t *off);
+#ifdef CONFIG_AFS_CACHEFS
 static int afs_file_page_mkwrite(struct page *page);
+#endif
 
 struct inode_operations afs_file_inode_operations = {
 	.getattr	= afs_inode_getattr,
@@ -56,7 +58,9 @@ struct address_space_operations afs_fs_a
 	.set_page_dirty	= __set_page_dirty_nobuffers,
 	.releasepage	= afs_file_releasepage,
 	.invalidatepage	= afs_file_invalidatepage,
+#ifdef CONFIG_AFS_CACHEFS
 	.page_mkwrite	= afs_file_page_mkwrite,
+#endif
 };
 
 /*****************************************************************************/
_
