Add a brief message describing running scheduler to dmesg.

Signed-off-by: Con Kolivas <kernel@kolivas.org>


Index: linux-2.6.10-rc1-mm4/init/main.c
===================================================================
--- linux-2.6.10-rc1-mm4.orig/init/main.c	2004-11-10 09:27:35.020244293 +1100
+++ linux-2.6.10-rc1-mm4/init/main.c	2004-11-10 09:28:07.207073503 +1100
@@ -47,6 +47,7 @@
 #include <linux/rmap.h>
 #include <linux/mempolicy.h>
 #include <linux/key.h>
+#include <linux/sched.h>
 
 #include <asm/io.h>
 #include <asm/bugs.h>
@@ -519,6 +520,7 @@ asmlinkage void __init start_kernel(void
 
 	acpi_early_init(); /* before LAPIC and SMP init */
 
+	printk("Running with %s cpu scheduler.\n", scheduler->cpusched_name);
 	/* Do the rest non-__init'ed, we're now alive */
 	rest_init();
 }
