

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/file.c |   10 +---------
 1 files changed, 1 insertion(+), 9 deletions(-)

diff -puN fs/file.c~file_tableexpand_files-code-cleanup-remove-debug fs/file.c
--- 25/fs/file.c~file_tableexpand_files-code-cleanup-remove-debug	Wed Jan 12 16:13:47 2005
+++ 25-akpm/fs/file.c	Wed Jan 12 16:14:07 2005
@@ -237,9 +237,6 @@ out:
 int expand_files(struct files_struct *files, int nr)
 {
 	int err, expand = 0;
-#ifdef FDSET_DEBUG
-	printk (KERN_ERR "%s %d: nr = %d\n", __FUNCTION__, current->pid, nr);
-#endif
 
 	if (nr >= files->max_fdset) {
 		expand = 1;
@@ -252,11 +249,6 @@ int expand_files(struct files_struct *fi
 			goto out;
 	}
 	err = expand;
- out:
-#ifdef FDSET_DEBUG
-	if (err)
-		printk (KERN_ERR "%s %d: return %d\n", __FUNCTION__, current->pid, err);
-#endif
+out:
 	return err;
 }
-
_
