
From: vlobanov <vlobanov@speakeasy.net>

Signed-off-by: Vadim Lobanov <vlobanov@speakeasy.net>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/i386/Kconfig |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN arch/i386/Kconfig~fix-typo-in-arch-i386-kconfig arch/i386/Kconfig
--- 25/arch/i386/Kconfig~fix-typo-in-arch-i386-kconfig	Wed Jan 12 15:49:51 2005
+++ 25-akpm/arch/i386/Kconfig	Wed Jan 12 15:49:51 2005
@@ -881,7 +881,7 @@ config REGPARM
 	depends on EXPERIMENTAL
 	default n
 	help
-	Compile the kernel with -mregparm=3. This uses an different ABI
+	Compile the kernel with -mregparm=3. This uses a different ABI
 	and passes the first three arguments of a function call in registers.
 	This will probably break binary only modules.
 
_
