

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/arch/x86_64/kernel/machine_kexec.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -L arch/i386/kernel/machine_kexec.c -puN /dev/null /dev/null
diff -puN arch/x86_64/kernel/machine_kexec.c~kexec-kexecx86_64-4level-fix arch/x86_64/kernel/machine_kexec.c
--- 25/arch/x86_64/kernel/machine_kexec.c~kexec-kexecx86_64-4level-fix	2004-11-04 00:16:47.160495088 -0800
+++ 25-akpm/arch/x86_64/kernel/machine_kexec.c	2004-11-04 00:18:59.000452368 -0800
@@ -218,7 +218,7 @@ void machine_kexec(struct kimage *image)
 	 * as that will happen when I fully switch to my identity mapped
 	 * page table anyway.
 	 */
-	memcpy((void *)read_pda(level4_pgt), __va(start_pgtable), PAGE_SIZE/2);
+	memcpy(current->active_mm->pml4, __va(start_pgtable), PAGE_SIZE/2);
 	__flush_tlb();
 
 
_
