

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/kernel/itimer.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletion(-)

diff -puN kernel/itimer.c~ltt-kernel-events-build-fix kernel/itimer.c
--- 25/kernel/itimer.c~ltt-kernel-events-build-fix	2005-01-13 23:01:12.813249240 -0800
+++ 25-akpm/kernel/itimer.c	2005-01-13 23:01:12.816248784 -0800
@@ -89,7 +89,9 @@ int do_setitimer(int which, struct itime
 
 	if (ovalue && (k = do_getitimer(which, ovalue)) < 0)
 		return k;
-	ltt_ev_timer(LTT_EV_TIMER_SETITIMER, which, i, j);
+	ltt_ev_timer(LTT_EV_TIMER_SETITIMER, which,
+			timeval_to_jiffies(&value->it_interval),
+			timeval_to_jiffies(&value->it_value));
 	switch (which) {
 		case ITIMER_REAL:
 			del_timer_sync(&current->real_timer);
_
