
From: Steve Longerbeam <stevel@mvista.com>


Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/fs/cachefs/block.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN fs/cachefs/block.c~numa-policies-for-file-mappings-mpol_mf_move-cachefs fs/cachefs/block.c
--- 25/fs/cachefs/block.c~numa-policies-for-file-mappings-mpol_mf_move-cachefs	Mon Nov 15 16:13:41 2004
+++ 25-akpm/fs/cachefs/block.c	Mon Nov 15 16:13:41 2004
@@ -374,7 +374,7 @@ int cachefs_block_cow(struct cachefs_sup
 		mapping = super->imisc->i_mapping;
 
 		ret = -ENOMEM;
-		newpage = page_cache_alloc_cold(mapping);
+		newpage = page_cache_alloc_cold(mapping, block->bix);
 		if (!newpage)
 			goto error;
 
_
