Index: xx-sources/arch/i386/Kconfig
===================================================================
--- xx-sources.orig/arch/i386/Kconfig	2004-11-18 23:30:24.000000000 +0000
+++ xx-sources/arch/i386/Kconfig	2005-01-14 10:00:35.372057600 +0000
@@ -1306,3 +1306,5 @@
 	bool
 	depends on X86 && !EMBEDDED
 	default y
+
+source "kernel/Kconfig-extra.xx"
Index: xx-sources/arch/x86_64/Kconfig
===================================================================
--- xx-sources.orig/arch/x86_64/Kconfig	2004-11-18 23:30:23.000000000 +0000
+++ xx-sources/arch/x86_64/Kconfig	2005-01-14 10:01:37.079676616 +0000
@@ -483,3 +483,5 @@
 source "crypto/Kconfig"
 
 source "lib/Kconfig"
+
+source "kernel/Kconfig-extra.xx"
Index: xx-sources/kernel/Kconfig-extra.xx
===================================================================
--- xx-sources.orig/kernel/Kconfig-extra.xx	2005-01-14 01:38:42.771666504 +0000
+++ xx-sources/kernel/Kconfig-extra.xx	2005-01-14 10:00:34.549182696 +0000
@@ -0,0 +1,4 @@
+menu "Experimental xx options"
+
+
+endmenu
