

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/filemap.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN mm/filemap.c~provide-a-filesystem-specific-syncable-page-bit-fix-2 mm/filemap.c
--- 25/mm/filemap.c~provide-a-filesystem-specific-syncable-page-bit-fix-2	2004-11-10 20:43:48.013683080 -0800
+++ 25-akpm/mm/filemap.c	2004-11-10 20:43:48.018682320 -0800
@@ -476,7 +476,7 @@ void fastcall end_page_fs_misc(struct pa
 	if (!TestClearPageFsMisc(page))
 		BUG();
 	smp_mb__after_clear_bit();
-	wake_up_page(page);
+	__wake_up_bit(page_waitqueue(page), &page->flags, PG_fs_misc);
 }
 
 EXPORT_SYMBOL(end_page_fs_misc);
_
