

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/filemap.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN mm/filemap.c~provide-a-filesystem-specific-syncable-page-bit-fix mm/filemap.c
--- 25/mm/filemap.c~provide-a-filesystem-specific-syncable-page-bit-fix	2004-11-10 20:43:47.805714696 -0800
+++ 25-akpm/mm/filemap.c	2004-11-10 20:43:47.809714088 -0800
@@ -476,7 +476,7 @@ void fastcall end_page_fs_misc(struct pa
 	if (!TestClearPageFsMisc(page))
 		BUG();
 	smp_mb__after_clear_bit();
-	wake_up_page(page, PG_fs_misc);
+	wake_up_page(page);
 }
 
 EXPORT_SYMBOL(end_page_fs_misc);
_
