

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 25-akpm/mm/filemap.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN mm/filemap.c~reiser4-export-find_get_pages_tag mm/filemap.c
--- 25/mm/filemap.c~reiser4-export-find_get_pages_tag	2005-01-10 21:15:24.116787272 -0800
+++ 25-akpm/mm/filemap.c	2005-01-10 21:15:24.121786512 -0800
@@ -635,7 +635,6 @@ unsigned find_get_pages(struct address_s
 	read_unlock_irq(&mapping->tree_lock);
 	return ret;
 }
-
 EXPORT_SYMBOL(find_get_pages);
 
 /*
@@ -658,6 +657,7 @@ unsigned find_get_pages_tag(struct addre
 	read_unlock_irq(&mapping->tree_lock);
 	return ret;
 }
+EXPORT_SYMBOL(find_get_pages_tag);
 
 /*
  * Same as grab_cache_page, but do not wait if the page is unavailable.
_
