Index: xx-sources/fs/proc/proc_misc.c
===================================================================
--- xx-sources.orig/fs/proc/proc_misc.c	2004-10-30 22:37:03.007026296 -0400
+++ xx-sources/fs/proc/proc_misc.c	2004-10-30 22:38:25.198531288 -0400
@@ -150,6 +150,52 @@
 	return proc_calc_metrics(page, start, off, count, eof, len);
 }
 
+static int xx_read_proc(char *page, char **start, off_t off,
+				int count, int *eof, void *data)
+{
+	const int jiffies = CONFIG_HERTZ;
+	const int idedelay = CONFIG_IDE_DELAY;
+	extern char *scheduler_name;
+	extern char *scheduler_version;
+#if defined(CONFIG_CFLAGS_EDIT)
+	const char *cflags = CONFIG_CFLAGS_STRING;
+#endif
+	int len;
+
+	len = sprintf(page,
+		"Process scheduler:			%s\n"
+		"Version:				%s\n"
+		"Supported I/O schedulers:\n"
+#if defined(CONFIG_IOSCHED_CFQ)
+		"					CFQ\n"
+#endif
+#if defined(CONFIG_IOSCHED_AS)
+		"					Anticipatory\n"
+#endif
+#if defined(CONFIG_IOSCHED_DEADLINE)
+		"					Deadline\n"
+#endif
+#if defined(CONFIG_IOSCHED_NOOP)
+		"					Noop\n"
+#endif
+		"Jiffies:				%d\n"
+		"IDE Delay:				%d\n"
+#if defined(CONFIG_CFLAGS_EDIT)
+		"CFLAGS:				%s\n"
+#endif
+		,
+		scheduler_name,
+		scheduler_version,
+		jiffies,
+		idedelay
+#if defined(CONFIG_CFLAGS_EDIT)
+			, cflags
+#endif
+		);
+
+	return proc_calc_metrics(page, start, off, count, eof, len);
+}
+
 static int meminfo_read_proc(char *page, char **start, off_t off,
 				 int count, int *eof, void *data)
 {
@@ -596,6 +642,7 @@
 		{"loadavg",     loadavg_read_proc},
 		{"uptime",	uptime_read_proc},
 		{"meminfo",	meminfo_read_proc},
+		{"xx",		xx_read_proc},
 		{"version",	version_read_proc},
 #ifdef CONFIG_PROC_HARDWARE
 		{"hardware",	hardware_read_proc},
