Index: xx-sources/fs/reiserfs/xattr.c
===================================================================
--- xx-sources.orig/fs/reiserfs/xattr.c	2004-11-04 15:23:07.488570544 -0500
+++ xx-sources/fs/reiserfs/xattr.c	2004-11-04 15:34:53.370260216 -0500
@@ -1410,7 +1410,9 @@
                 }
 #endif
 	} else {
+#ifdef CONFIG_REISERFS_FS_POSIX_ACL
 check_groups:
+#endif
 		if (in_group_p(inode->i_gid))
 			mode >>= 3;
 	}
@@ -1421,7 +1423,9 @@
 	if (((mode & mask & (MAY_READ|MAY_WRITE|MAY_EXEC)) == mask))
 		return 0;
 
+#ifdef CONFIG_REISERFS_FS_POSIX_ACL
 check_capabilities:
+#endif
 	/*
 	 * Read/write DACs are always overridable.
 	 * Executable DACs are overridable if at least one exec bit is set.
